File: pt5/1129.shortest-path-with-alternating-colors.java
import java.util.*;

class Solution {
    @SuppressWarnings("unchecked")
    public int[] shortestAlternatingPaths(int n, int[][] redEdges, int[][] blueEdges) {
        List<Integer>[][] graph = new ArrayList[2][n];
        for (int color = 0; color < 2; color++) {
            for (int node = 0; node < n; node++) {
                graph[color][node] = new ArrayList<>();
            }
        }

        for (int[] edge : redEdges) {
            graph[0][edge[0]].add(edge[1]);
        }
        for (int[] edge : blueEdges) {
            graph[1][edge[0]].add(edge[1]);
        }

        int[] answer = new int[n];
        Arrays.fill(answer, -1);
        boolean[][] visited = new boolean[n][2];
        Queue<int[]> queue = new ArrayDeque<>();
        queue.offer(new int[] {0, 0, 0});
        queue.offer(new int[] {0, 1, 0});
        visited[0][0] = true;
        visited[0][1] = true;

        while (!queue.isEmpty()) {
            int[] state = queue.poll();
            int node = state[0];
            int lastColor = state[1];
            int distance = state[2];

            if (answer[node] == -1) {
                answer[node] = distance;
            }

            int nextColor = 1 - lastColor;
            for (int neighbor : graph[nextColor][node]) {
                if (!visited[neighbor][nextColor]) {
                    visited[neighbor][nextColor] = true;
                    queue.offer(new int[] {neighbor, nextColor, distance + 1});
                }
            }
        }

        return answer;
    }
}

File: pt5/1129.shortest-path-with-alternating-colors.go
package main

func shortestAlternatingPaths(n int, redEdges [][]int, blueEdges [][]int) []int {
	graph := make([][][]int, 2)
	graph[0] = make([][]int, n)
	graph[1] = make([][]int, n)

	for _, edge := range redEdges {
		graph[0][edge[0]] = append(graph[0][edge[0]], edge[1])
	}
	for _, edge := range blueEdges {
		graph[1][edge[0]] = append(graph[1][edge[0]], edge[1])
	}

	answer := make([]int, n)
	for i := range answer {
		answer[i] = -1
	}

	type state struct {
		node      int
		lastColor int
		distance  int
	}

	visited := make([][2]bool, n)
	queue := []state{{node: 0, lastColor: 0}, {node: 0, lastColor: 1}}
	visited[0][0], visited[0][1] = true, true

	for head := 0; head < len(queue); head++ {
		current := queue[head]
		if answer[current.node] == -1 {
			answer[current.node] = current.distance
		}

		nextColor := 1 - current.lastColor
		for _, neighbor := range graph[nextColor][current.node] {
			if !visited[neighbor][nextColor] {
				visited[neighbor][nextColor] = true
				queue = append(queue, state{node: neighbor, lastColor: nextColor, distance: current.distance + 1})
			}
		}
	}

	return answer
}

